File: leetcode/src/palindrome-pairs.py
# https://leetcode.com/problems/palindrome-pairs/description/
# O(kn^2) solution
def get_prefixes(word: str) -> List[str]:
    prefixes = []
    for left in range(len(word)):
        if word[left:] == word[left:][::-1]:
            prefixes.append(word[:left]) 
    return prefixes

def get_suffixes(word: str) -> List[str]:
    suffixes = []
    for right in range(1, len(word)+1):
        if word[:right] == word[:right][::-1]:
            suffixes.append(word[right:]) 
    return suffixes


class Solution:
    def palindromePairs(self, words: List[str]) -> List[List[int]]:
        word_reverse_mapping = {word:i for i, word in enumerate(words)}
        solutions = []

        for i, word in enumerate(words):
            reversed_word = word[::-1]
            if (reversed_word in word_reverse_mapping and word_reverse_mapping[reversed_word] != i):
                solutions.append([i, word_reverse_mapping[reversed_word]])

            for substr in get_prefixes(word):
                reversed_word = substr[::-1]
                if (reversed_word in word_reverse_mapping and word_reverse_mapping[reversed_word] != i):
                    solutions.append([i, word_reverse_mapping[reversed_word]])

            for substr in get_suffixes(word):
                reversed_word = substr[::-1]
                if (reversed_word in word_reverse_mapping and word_reverse_mapping[reversed_word] != i):
                    solutions.append([word_reverse_mapping[reversed_word], i])
            
        return solutions

            

File: leetcode/src/palindrome-pairs.cpp
// https://leetcode.com/problems/palindrome-pairs/description/
// O(kn^2) solution

#include <unordered_map>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<string> getPrefixes(const string& word) {
        vector<string> result;
        for (int i = word.size()-1; i >= 0; i --) {

            int lo = i, hi = word.size()-1;
            bool isPalindrome = true;
            while (lo < hi) {
                if (word[lo] != word[hi]) {
                    isPalindrome = false;
                    break;
                }
                lo ++; hi --;
            }

            if (isPalindrome) {
                result.push_back(word.substr(0,i));
            }

        }

        return result;
    }

    vector<string> getSuffixes(const string& word) {
        vector<string> result;

        for (int i = 1; i <= word.size(); i ++) {
            int lo = 0, hi = i-1;
            bool isPalindrome = true;
            while (lo < hi) {
                if (word[lo] != word[hi]) {
                    isPalindrome = false;
                    break;
                }
                lo ++; hi --;
            }
            if (isPalindrome) {
                result.push_back(word.substr(i, word.size()));
            }
        }

        return result;
    }

    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> result;

        unordered_map<string, int> reverse_mapping;
        for (int i = 0; i < words.size(); i ++) {
            reverse_mapping[words[i]] = i;
        }

        for (int i = 0; i < words.size(); i ++) {
            const string& word = words[i];
            string reversed_word (word.rbegin(), word.rend());
            if (reverse_mapping.find(reversed_word) != reverse_mapping.end() && reverse_mapping[reversed_word] != i) {
                result.push_back({i, reverse_mapping[reversed_word]});
            }

            for (const string& prefix : getPrefixes(word)) {
                string reversed_word (prefix.rbegin(), prefix.rend());
                if (reverse_mapping.find(reversed_word) != reverse_mapping.end() && reverse_mapping[reversed_word] != i) {
                    result.push_back({i, reverse_mapping[reversed_word]});
                }
            }

            for (const string& suffix: getSuffixes(word)) {
                string reversed_word (suffix.rbegin(), suffix.rend());
                if (reverse_mapping.find(reversed_word) != reverse_mapping.end() && reverse_mapping[reversed_word] != i) {
                    result.push_back({reverse_mapping[reversed_word],i});
                }
            }
        }


        


        return result;
    }
};